### PR TITLE
[TEVA 1684] Dashboard slim shady view component

### PR DESCRIPTION
**Pull request by alexbowen**
https://dfedigital.atlassian.net/browse/TEVA-1625

implementation of a unified dashboard view component that aligns all current dashboard instances in a visual and functional way - jobseeker account and MAT dashboard. there will be several more instances over the coming months of other dashboards.

Screenshot 2021-01-06 at 15 34 19

Screenshot 2021-01-06 at 15 34 00

Screenshot 2021-01-06 at 15 33 42

Screenshot 2021-01-06 at 15 51 14

### CODE REVIEW

**Comment by martin-bangoura** [comment]
Review app deployed to <https://teaching-vacancies-review-pr-2597.london.cloudapps.digital>

File: app/components/publishers/vacancies_component.rb
Comment (cpjmcquillan) [inline app/components/publishers/vacancies_component.rb:26]: What's the second part of this line trying to check?

File: app/components/publishers/vacancies_component/vacancies_component.html.slim
Comment (cpjmcquillan) [inline app/components/publishers/vacancies_component/vacancies_component.html.slim:30]: the `method: "patch"` can be outside the html hash and just be `method: :patch` like it is on line 12

File: app/components/shared/dashboard_component/dashboard_component.html.slim
Comment (cpjmcquillan) [inline app/components/shared/dashboard_component/dashboard_component.html.slim:9]: This has changed in https://github.com/DFE-Digital/teaching-vacancies/pull/2598

Comment (cpjmcquillan) [reply app/components/shared/dashboard_component/dashboard_component.html.slim:9]: Also, if we're passing in the link, should we pass in the link text too so it isn't tied to `t("buttons.create_job")`?

File: app/views/jobseekers/application/_account_header.html.haml
Comment (cpjmcquillan) [inline app/views/jobseekers/application/_account_header.html.haml:1]: You don't need to pass in `link: false`, you can just omit it, and then when you do `if @link` maybe do something like `if link.present`

File: app/components/publishers/vacancies_component.rb
Comment (alexbowen) [reply app/components/publishers/vacancies_component.rb:26]: there is no sorting for awaiting feedback page and hence there are no options for this selected_type

Comment (alexbowen) [reply app/components/publishers/vacancies_component.rb]: have changed this to be a `.present?` check

File: app/components/shared/dashboard_component.rb
Comment (cpjmcquillan) [inline app/components/shared/dashboard_component.rb:4]: I'd probably do `link: nil`

File: app/components/shared/dashboard_component.rb
Comment (cpjmcquillan) [inline app/components/shared/dashboard_component.rb:2]: since we just read them, we can probably do `attr_reader`

File: app/components/shared/dashboard_component/dashboard_component.html.slim
Comment (cpjmcquillan) [inline app/components/shared/dashboard_component/dashboard_component.html.slim:9]: When we do `attr_accessor` or `attr_reader` you can just access the attribute by name, you don't need to access the instance variable. (Same applies to line 4 and line 14)

Comment (cpjmcquillan) [reply app/components/shared/dashboard_component/dashboard_component.html.slim:9]: Also, `method` is `:post` by default, so no need to specify that for `govuk_button_to`

File: spec/system/hiring_staff_can_sign_in_with_dfe_spec.rb
Comment (david-mears-dfe) [inline spec/system/hiring_staff_can_sign_in_with_dfe_spec.rb:106]: All these should be `eq` instead of `eql`: https://github.com/DFE-Digital/teaching-vacancies/pull/2585

File: spec/system/hiring_staff_can_add_stats_to_a_vacancy_spec.rb
Comment (david-mears-dfe) [inline spec/system/hiring_staff_can_add_stats_to_a_vacancy_spec.rb:30]: Are we removing the 'awaiting feedback' notifications?
If so, please can you remove the other two instances of `NOTIFICATION_BADGE_SELECTOR`

File: spec/system/hiring_staff_can_sign_in_with_dfe_spec.rb
Comment (alexbowen) [reply spec/system/hiring_staff_can_sign_in_with_dfe_spec.rb:106]: good spot, dodgy rebase on my part

File: spec/system/hiring_staff_can_add_stats_to_a_vacancy_spec.rb
Comment (alexbowen) [reply spec/system/hiring_staff_can_add_stats_to_a_vacancy_spec.rb:30]: will address this, but its imminent we will be removing the awaiting feedback stuff altogether

Comment (alexbowen) [reply spec/system/hiring_staff_can_add_stats_to_a_vacancy_spec.rb:30]: when chris gives the thumbs up, ticket in backlog already

File: app/components/publishers/vacancies_component/vacancies_component.html.slim
Comment (david-mears-dfe) [inline app/components/publishers/vacancies_component/vacancies_component.html.slim:6]: You might be able to use `==`. Cesi says

> Tip about slim, when you use the double == you do not need to have the translation key ending with _html

Conversation:

https://ukgovernmentdfe.slack.com/archives/CQ3UX00P6/p1610034056113300?thread_ts=1610033463.108900&cid=CQ3UX00P6

File: app/components/publishers/vacancies_component.rb
Comment (david-mears-dfe) [inline app/components/publishers/vacancies_component.rb:21]: It is Rubyish to make Boolean methods end in `?`

File: app/components/publishers/vacancies_component.rb
Comment (david-mears-dfe) [inline app/components/publishers/vacancies_component.rb:49]: Though I think you can't put a `?` in a parameter name

File: app/components/publishers/vacancies_component.rb
Comment (david-mears-dfe) [inline app/components/publishers/vacancies_component.rb:25]: `sortable?`

File: app/components/publishers/vacancies_component.rb
Comment (david-mears-dfe) [reply app/components/publishers/vacancies_component.rb:49]: Is there a reason to pass the `selected` as a parameter rather than directly calling the `selected` method within this method `vacancy_type_tab_link`?

File: app/components/publishers/school_overview_component.html.slim
Comment (david-mears-dfe) [inline app/components/publishers/school_overview_component.html.slim:41]: Perhaps instead of `if/elsif`:

```ruby
- if (@organisation.website || @organisation.url).present?
= govuk_link_to((@organisation.website || @organisation.url), (@organisation.website || @organisation.url), class: "wordwrap", target: "_blank")
```

Or add a helper to OrganisationHelper that gets `(@organisation.website || @organisation.url)`?

Comment (david-mears-dfe) [reply app/components/publishers/school_overview_component.html.slim:41]: I realise you are just trying to remove an indentation, not really adding these lines...

File: app/models/organisation.rb
Comment (david-mears-dfe) [inline app/models/organisation.rb:13]: How about this (but with better naming):

```ruby
JOB_SORTING_OPTIONS_BASE = [
    [I18n.t("jobs.sort_by.expires_on.ascending"), "expires_on"],
    [I18n.t("jobs.sort_by.job_title.ascending"), "job_title"],
    [I18n.t("jobs.sort_by.location.ascending"), "readable_job_location"]
].freeze

JOB_SORTING_OPTIONS_EXTENDED = (JOB_SORTING_OPTIONS_BASE + [I18n.t("jobs.sort_by.published_date.ascending"), "publish_on"]).freeze

JOB_SORTING_OPTIONS = {
    published: JOB_SORTING_OPTIONS_BASE,
    pending: JOB_SORTING_OPTIONS_EXTENDED,
    draft: JOB_SORTING_OPTIONS_EXTENDED,
    expired: JOB_SORTING_OPTIONS_BASE,
  }.freeze
```

File: config/locales/jobs.yml
Comment (david-mears-dfe) [inline config/locales/jobs.yml:103]: Some of the sort-bys use title case and others are all lower case. Could we align on one or the other (in consultation with @RoseMackworth)?

Either way, I think (Oldest) and (Soonest) will want to be lower case.

Comment (alexbowen) [reply config/locales/jobs.yml:103]: yeah well picked up, i actually changed them to lowercase and changed them back again to align with figma and less friction (maybe laziness) biut you are right it should be one or the other and consistent

File: app/components/publishers/vacancies_component.rb
Comment (alexbowen) [reply app/components/publishers/vacancies_component.rb:25]: it is really is_sortable but i got picked up on this before that it is not the ruby convention to prepend with is_ ???

**Comment by david-mears-dfe** [comment]
Screenshot 2021-01-07 at 17 29 30

I think this should be 'Job filters'? Check with @RoseMackworth please

File: app/components/publishers/vacancies_component.rb
Comment (alexbowen) [reply app/components/publishers/vacancies_component.rb:25]: oh no sorry i see your point, its boolean and needs ?

**Comment by david-mears-dfe** [comment]
Screenshot 2021-01-07 at 17 32 00
At least on my browser (Chrome) the magnifying glass icon is cut off at the top and side.

File: app/components/publishers/vacancies_component/vacancies_component.html.slim
Comment (alexbowen) [reply app/components/publishers/vacancies_component/vacancies_component.html.slim:6]: yeah i saw this thread earlier and intended to align with this

**Comment by alexbowen** [comment]
> Screenshot 2021-01-07 at 17 29 30
> 
> I think this should be 'Job filters'? Check with @RoseMackworth please

ok, will check, but these kind of things is getting a bit out of scope as i used exact same copy that already existed

**Comment by david-mears-dfe** [comment]
(On mobile)
These should look more alike, right?
Screenshot 2021-01-07 at 17 34 59
Screenshot 2021-01-07 at 17 35 04
Screenshot 2021-01-07 at 17 35 08

**Comment by alexbowen** [comment]
> Screenshot 2021-01-07 at 17 32 00
> At least on my browser (Chrome) the magnifying glass icon is cut off at the top and side.

eagle eyes, sounds like you need a day off ;)

**Comment by david-mears-dfe** [comment]
Screenshot 2021-01-07 at 17 40 08

I may be out of scope here. Should this bit display another dotted line box that says "You currently have no saved jobs"? [Figma](https://www.figma.com/file/IB5MUgCC5Kf084XxafUjG1/Sprint-66?node-id=0%3A1) doesn't seem to plan for the no-saved-jobs scenario.

File: config/locales/jobs.yml
Comment (alexbowen) [reply config/locales/jobs.yml:103]: confirmed everything should be lowercase

**Review by david-mears-dfe — APPROVED** [review]
Approved if you squash 'review comment updates' before merging.

File: app/controllers/publishers/organisations/managed_organisations_controller.rb
Comment (cpjmcquillan) [inline app/controllers/publishers/organisations/managed_organisations_controller.rb:12]: This controller is only for the managed organisations page, so sorting here is never applicable

File: app/controllers/publishers/organisations/managed_organisations_controller.rb
Comment (cpjmcquillan) [inline app/controllers/publishers/organisations/managed_organisations_controller.rb:29]: same as above

File: app/form_models/managed_organisations_form.rb
Comment (cpjmcquillan) [inline app/form_models/managed_organisations_form.rb:11]: I'd probably prefer a form just for sorting so as not to conflate the two things that really don't need to be related. But if we don't do that, then `sort: nil` as the second parameter would allow you to only specify the sort when it's relevant

File: app/frontend/src/styles/application/form.scss
Comment (cpjmcquillan) [inline app/frontend/src/styles/application/form.scss:27]: nice 👍

File: app/form_models/managed_organisations_form.rb
Comment (cpjmcquillan) [inline app/form_models/managed_organisations_form.rb:7]: You can lose all this from here now

File: app/form_models/vacancies_sort_form.rb
Comment (cpjmcquillan) [inline app/form_models/vacancies_sort_form.rb:1]: `VacancySortForm` I _think_ is a more conventional name

File: app/form_models/vacancy_sort_form.rb
Comment (cpjmcquillan) [inline app/form_models/vacancy_sort_form.rb:6]: You can just pass the param like normal if it's always passed

**Review by cpjmcquillan — APPROVED** [review]

**Pull request merged**
